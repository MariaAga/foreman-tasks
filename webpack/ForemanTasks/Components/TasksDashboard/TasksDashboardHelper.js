import URI from 'urijs';
import { getURIQuery } from 'foremanReact/common/helpers';

import {
  TASKS_DASHBOARD_AVAILABLE_TIMES,
  TASKS_DASHBOARD_QUERY_KEYS_TEXT,
  TASKS_DASHBOARD_QUERY_VALUES_TEXT,
} from './TasksDashboardConstants';

export const getQueryKeyText = key => TASKS_DASHBOARD_QUERY_KEYS_TEXT[key];

export const getQueryValueText = value =>
  TASKS_DASHBOARD_QUERY_VALUES_TEXT[value];

export const timeToHoursNumber = time => {
  switch (time) {
    case TASKS_DASHBOARD_AVAILABLE_TIMES.H12:
      return 12;
    case TASKS_DASHBOARD_AVAILABLE_TIMES.H24:
      return 24;
    case TASKS_DASHBOARD_AVAILABLE_TIMES.WEEK:
      return 24 * 7;
    default:
      return 24;
  }
};

const uriToQueryMap = {
  state: 'state',
  result: 'result',
  time_mode: 'mode',
  time_horizon: 'time',
};

const queryFromUriQuery = uriQuery => {
  const query = {};

  Object.entries(uriToQueryMap).forEach(([uriField, queryField]) => {
    if (uriQuery[uriField]) query[queryField] = uriQuery[uriField];
  });

  if (query.mode === 'recent') {
    query.mode = 'last';
  }

  return query;
};

const uriQueryFromQuery = query => {
  const uriQuery = {};

  Object.entries(uriToQueryMap).forEach(([uriField, queryField]) => {
    if (query[queryField]) uriQuery[uriField] = query[queryField];
  });

  if (uriQuery.time_mode === 'last') {
    uriQuery.time_mode = 'recent';
  }

  return uriQuery;
};

export const getQueryFromUrl = () => {
  const uriQuery = getURIQuery(window.location.href);

  return queryFromUriQuery(uriQuery);
};

export const updateURLQuery = ({ state, result, mode, time, uri }) => {
  const query = {
    state,
    result,
    time_mode: mode === 'last' ? 'recent' : mode,
    time_horizon: time,
    page: 1,
  };

  uri.setSearch(query);
  window.history.pushState({ path: uri.toString() }, '', uri.toString());
};

export const resolveQuery = query => {
  const uriQuery = uriQueryFromQuery(query);

  const uri = new URI(window.location.href);
  const { search, per_page: perPage } = uri.query(true);

  const data = { search, ...uriQuery, page: 1, per_page: perPage };
  const { $, tfm } = window;

  updateURLQuery({ uri, ...query });

  uri.query(URI.buildQuery(data, true));
  tfm.tools.showSpinner();
  $.ajax({
    type: 'get',
    url: uri.toString(),
    success(result) {
      const res = $(`<div>${result}</div>`);

      $('#tasks-table').html(res.find('#tasks-table'));
    },
    error({ statusText }) {
      $('#tasks-table').html(statusText);
    },
    complete(result) {
      tfm.tools.hideSpinner();
    },
  });
};
