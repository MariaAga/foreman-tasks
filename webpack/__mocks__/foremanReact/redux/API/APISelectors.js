export const selectAPIResponse = (state, key) => ({
  data: {
    text: 'some-data',
    key,
  },
});

export const selectAPIStatus = (state, key) => 'PENDING';
export const selectAPIError = (state, key) => ({ error: `${key} ERRROR` });
