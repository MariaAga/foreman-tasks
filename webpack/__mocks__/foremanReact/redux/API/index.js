export const API = {
  get: jest.fn(),
  put: jest.fn(),
  post: jest.fn(),
  delete: jest.fn(),
  patch: jest.fn(),
};

export const get = data => ({ type: 'some-type', ...data });
