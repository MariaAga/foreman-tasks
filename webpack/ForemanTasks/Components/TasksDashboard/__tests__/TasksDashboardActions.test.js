import { testActionSnapshotWithFixtures } from '@theforeman/test';
import { API } from 'foremanReact/redux/API';
import { timeToHoursNumber } from '../TasksDashboardHelper';
import {
  initializeDashboard,
  updateTime,
  updateQuery,
  fetchTasksSummary,
} from '../TasksDashboardActions';
import {
  correctTime,
  wrongTime,
  parentTaskID,
  apiGetMock,
} from './TaskDashboard.fixtures';

jest.mock('foremanReact/redux/API');
jest.mock('../TasksDashboardHelper');

timeToHoursNumber.mockImplementation(arg => arg);
API.get.mockImplementation(apiGetMock);

const fixtures = {
  'should initialize-dashboard': () =>
    initializeDashboard({ time: 'some-time', query: 'some-query' }),
  'should update-time': () => updateTime('some-time'),
  'should update-query': () => updateQuery('some-query'),
  'should fetch-tasks-summary and success': () =>
    fetchTasksSummary(correctTime),
  'should fetch-tasks-summary for subtasks and success': () =>
    fetchTasksSummary(correctTime, parentTaskID),
  'should fetch-tasks-summary and fail': () => fetchTasksSummary(wrongTime),
};

describe('TasksDashboard - Actions', () =>
  testActionSnapshotWithFixtures(fixtures));
