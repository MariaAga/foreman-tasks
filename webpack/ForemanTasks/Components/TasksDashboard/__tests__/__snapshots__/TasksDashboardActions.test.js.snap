// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksDashboard - Actions should fetch-tasks-summary and fail 1`] = `
Array [
  Array [
    Object {
      "type": "FOREMAN_TASKS_DASHBOARD_FETCH_TASKS_SUMMARY_REQUEST",
    },
  ],
  Array [
    Object {
      "payload": [Error: wrong time],
      "type": "FOREMAN_TASKS_DASHBOARD_FETCH_TASKS_SUMMARY_FAILURE",
    },
  ],
]
`;

exports[`TasksDashboard - Actions should fetch-tasks-summary and success 1`] = `
Array [
  Array [
    Object {
      "type": "FOREMAN_TASKS_DASHBOARD_FETCH_TASKS_SUMMARY_REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "tasksSummary": Object {
          "paused": Object {
            "recent": 0,
            "total": 54,
          },
          "running": Object {
            "recent": 0,
            "total": 3,
          },
          "scheduled": Object {
            "recent": 0,
            "total": 7,
          },
          "stopped": Object {
            "by_result": Object {
              "cancelled": Object {
                "recent": 0,
                "total": 105,
              },
              "error": Object {
                "recent": 0,
                "total": 121,
              },
              "success": Object {
                "recent": 0,
                "total": 532,
              },
              "warning": Object {
                "recent": 0,
                "total": 10,
              },
            },
            "recent": 0,
            "total": 768,
          },
        },
      },
      "type": "FOREMAN_TASKS_DASHBOARD_FETCH_TASKS_SUMMARY_SUCCESS",
    },
  ],
]
`;

exports[`TasksDashboard - Actions should fetch-tasks-summary for subtasks and success 1`] = `
Array [
  Array [
    Object {
      "type": "FOREMAN_TASKS_DASHBOARD_FETCH_TASKS_SUMMARY_REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "tasksSummary": Object {
          "paused": Object {
            "recent": 0,
            "total": 7,
          },
          "running": Object {
            "recent": 0,
            "total": 3,
          },
          "scheduled": Object {
            "recent": 0,
            "total": 7,
          },
          "stopped": Object {
            "by_result": Object {
              "cancelled": Object {
                "recent": 0,
                "total": 0,
              },
              "error": Object {
                "recent": 0,
                "total": 0,
              },
              "success": Object {
                "recent": 0,
                "total": 0,
              },
              "warning": Object {
                "recent": 0,
                "total": 0,
              },
            },
            "recent": 0,
            "total": 0,
          },
        },
      },
      "type": "FOREMAN_TASKS_DASHBOARD_FETCH_TASKS_SUMMARY_SUCCESS",
    },
  ],
]
`;

exports[`TasksDashboard - Actions should initialize-dashboard 1`] = `
Object {
  "payload": Object {
    "query": "some-query",
    "time": "some-time",
  },
  "type": "FOREMAN_TASKS_DASHBOARD_INIT",
}
`;

exports[`TasksDashboard - Actions should update-query 1`] = `
Array [
  Array [
    Object {
      "payload": "some-query",
      "type": "FOREMAN_TASKS_DASHBOARD_UPDATE_QUERY",
    },
  ],
]
`;

exports[`TasksDashboard - Actions should update-time 1`] = `
Object {
  "payload": "some-time",
  "type": "FOREMAN_TASKS_DASHBOARD_UPDATE_TIME",
}
`;
